File: solutions/haiku-1/haiku/ruby/wordfreq.rb
# Read all input
input = STDIN.read

# Convert to lowercase for case-insensitive comparison
input = input.downcase

# Extract words (maximal runs of ASCII letters) and count them
word_count = {}
input.scan(/[a-z]+/) do |word|
  word_count[word] = (word_count[word] || 0) + 1
end

# Sort: by count descending, then by word ascending
sorted = word_count.sort_by { |word, count| [-count, word] }

# Output
sorted.each do |word, count|
  puts "#{word} #{count}"
end

File: solutions/haiku-1/haiku/python/wordfreq.py
import sys
import re
from collections import Counter

# Read all input
text = sys.stdin.read()

# Extract words (maximal runs of ASCII letters)
words = re.findall(r'[a-zA-Z]+', text)

# Convert to lowercase and count
word_counts = Counter(word.lower() for word in words)

# Sort by count (descending), then by word (ascending)
sorted_words = sorted(word_counts.items(), key=lambda x: (-x[1], x[0]))

# Output
for word, count in sorted_words:
    print(f"{word} {count}")
